### ISSUE TITLE
(bad?) ANSI escape sequences cause what appears to be a random newline in output

### ISSUE THREAD

**Issue by pmorch** [bats-core/bats-core#480]
This bats file:

```bats
@test "test1" {
    echo "Some output" >&3
}

@test "test2" {
}
```

Prints output that looks like this in the terminal:

```bash
WSL2@~» bats ansi.bats 
 ✓ test1
Some output

 ✓ test2

2 tests, 0 failures

WSL2@~» 
```
Notice the unexpected newline after "Some output".

The unexpected newline is there the first 3-10 times I run it. After a couple of runs, the extra newline after "Some output" disappears and I see this - expected - output:

```bash
WSL2@~» bats ansi.bats
 ✓ test1
Some output
 ✓ test2

2 tests, 0 failures

WSL2@~» 
```

It stays like that (without the extra newline) for the next 100 runs. Until I run `reset` to clear the screen. Then it prints the output with the newline "a couple of times", and then the version without newlines forever after.

It gets weirder.

The output from `bats` actually remains the same for every run. If I do:

```bash
bats --pretty ansi.bats  > file-with-ansi
```
and then `cat file-with-ansi` over and over, then the first few times it has an extra newline, but after a while, the extra newline disappears. 

Yup. I kid you not. Try it yourself.

I've attached [file-with-ansi.zip](https://github.com/bats-core/bats-core/files/6978842/file-with-ansi.zip) containing only `file-with-ansi` that on my machine running `bats --version == Bats 1.4.1` has a sha256sum of `bcd5650aca4cb84cf870a1dc31d0cfa8cc96f0f9e5a371bd7946b2b7c0584d39`

I've reproduced this with `xfce4-terminal`, `xterm` and the Windows terminal under WSL2, and with `xfce4-terminal` and `gnome-terminal` on a "real" Ubuntu 18.04. The odd thing, is that the number of `cat file-with-ansi` that have the unexpected newline is constant for each terminal, but the constant varies from terminal to terminal. E.g. when running:

```bash
clear ; for i in $(seq 1 10) ; do echo Iteration: $i ; cat file-with-ansi ; done
````

Windows terminal shows the extra newline the first 4 iterations. For `xfce4-terminal` it shows 8 iterations with extra newline. 
For `xterm` it is 20.

By no stretch of the imagination am I an "ANSI Escape Code" expert. But I tried to decode the ANSI escape codes with this simple perl script based on this [gist](https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797). It is not complete, it can just decode the ANSI escape sequences in `file-with-ansi` from my above test case. Here is what (I think) the output of file can be decoded to:

```
WSL2@~» ./idea/lsc/parseAnsiEscape.pl file-with-ansi
Move cursor to column 1
Normal chars: "   test1"
Clear current line
Move cursor to column 77
Normal chars: "1/2"
Move cursor to column 2
line done
Normal chars: "Some output"
line done
Move cursor up 2 lines
Move cursor to column 1
Normal chars: " ✓ test1"
Clear current line
line done
Mode: reset all modes
Move cursor down 2 lines
Move cursor to column 1
Normal chars: "   test2"
Clear current line
Move cursor to column 77
Normal chars: "2/2"
Move cursor to column 2
Move cursor to column 1
Normal chars: " ✓ test2"
Clear current line
line done
Mode: reset all modes
line done
Normal chars: "2 tests, 0 failures"
line done
line done
```

I can't personally identify anything wrong with the `bats` output. But the output inconsistency speaks for itself. :-(

Here is my very incomplete and likely buggy [ansi decoder](https://gist.github.com/pmorch/157630b2de838c29ea14b5a62f263931#file-parseansiescape-pl) used to produce the output above.

**Comment by martin-schulze-vireso** [comment]
Wow that is really strange. I cannot reproduce this issue on my local (gnome?) terminal. However, this:

```
Normal chars: " ✓ test1"
Clear current line
line done
Mode: reset all modes
Move cursor down 2 lines
```

Looks wrong, going to the next line (`line done`) after printing test and then moving down 2 lines should leave one line empty. Yet, there is no gap on my system.

I tried a fix for that: change `libexec/bats-core/bats-format-pretty` to `move_down $(( line_backoff_count - 1 ))` and see if the gap is gone. I find it rather curious that applying this change did not change the output on my shell.

**Comment by martin-schulze-vireso** [comment]
I tried installing xterm but could not reproduce the issue there. Did you try exclusively on Windows?

**Comment by pmorch** [comment]
I think your hypothesis is correct. Whenever there is "room" below, then it adds a newline but when it reaches the bottom of a window it looks good.

Very odd. I've tried this on a Windows work-laptop running 1.4.1 from brew *and* on my own Linux desktop using brew 1.4.1 and git master 7ff2f3efc738976feaccfdf374164b61599f9e36 . And I did a clean-room maneuver where on my Linux desktop I cut'n'pasted from this issue. No scp-ing of files to make sure weird chars in files wasn't the reason. 

The file produced by my example that one can `cat` is also in [file-with-ansi.zip](https://github.com/bats-core/bats-core/files/6978842/file-with-ansi.zip).
